pkg/kubelet/cm: cgroup-related cleanups

kolyshkin opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

While working on #102147, I noticed a few odd things in the code that can be straightened out.
This PR does some of that. Please see detailed description in commits.

#### Which issue(s) this PR fixes:

none

#### Special notes for your reviewer:

Please review on commit by commit basis.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

none

kolyshkin committed:
kubelet/cm: rm propagateControllers

This was added by commit a9772b2290b4.

In the current codebase, the cgroup being updated was created using
runc/opencontainers' manager.Apply(), which already does controllers
propagation, so there is no need to repeat that on every update.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

ehashman commented:
/ok-to-test
/priority backlog
/triage accepted
/cc @odinuge @fromanirh @klueska

kolyshkin committed:
kubelet/cm: GetResourceStats -> MemoryUsage

Commit cc50aa9dfb2 introduced GetResourceStats, a method which collected
all the statistics from various cgroup controllers, only to discard all
of the info collected except a single value (memory usage).

While one may argue that this method can potentially be used from other
places, this did not happen since it was added 4+ years ago.

Let's streamline this code and only collect what we need, i.e. memory
usage. Rename the method accordingly.

While at it, fix pkg/kubelet/cm/cgroup_manager_unsupported.go to not
instantiate a new error every time a method is called.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

fromanirh commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -395,15 +394,6 @@ func getSupportedUnifiedControllers() sets.String {
-				Major:       libcontainerdevices.Wildcard,
-			},
-		},
 		SkipDevices: true,
```
this intuitively makes sense to me, but could you please add a reference to the specs/docs (a line in the git commit is fine like `see: https://....`) for future reference?

kolyshkin replied:
I was going to refer to https://github.com/opencontainers/runc/pull/2490, but unfortunately it falls short in case of systemd -- one more PR is needed (created: https://github.com/opencontainers/runc/pull/2958).

Removing this commit until https://github.com/opencontainers/runc/pull/2958 is merged, a release is cut, and kubernetes vendoring is updated to include it.

odinuge commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -393,62 +373,8 @@ func getSupportedUnifiedControllers() sets.String {
-
 func (m *cgroupManagerImpl) toResources(resourceConfig *ResourceConfig) *libcontainerconfigs.Resources {
 	resources := &libcontainerconfigs.Resources{
-		Devices: []*libcontainerdevices.Rule{
```
Have you tested this with systemd? I thought this was required for that, but maybe that has changed?

kolyshkin replied:
As explained above (https://github.com/kubernetes/kubernetes/pull/102218#discussion_r637454215), I rushed to make this change. Removing this commit until https://github.com/opencontainers/runc/pull/2958 is merged, updated runc is released and vendored in here.

odinuge commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -669,53 +591,21 @@ func (m *cgroupManagerImpl) ReduceCPULimits(cgroupName CgroupName) error {
-			return nil, fmt.Errorf("failed to get stats for cgroup %v: %v", name, err)
-		}
+		path = m.buildCgroupUnifiedPath(name)
+		file = "memory.max"
```
Is this correct? Shouldn't this be `memory.current`?

kolyshkin replied:
🤦🏻 my bad; fixed

odinuge reviewed (changes requested):
Overall change is ok, but still has some potential to break stuff. This is not urgent, so I think we should wait for https://github.com/kubernetes/kubernetes/issues/102148 and possibly https://github.com/kubernetes/test-infra/pull/22290.

This should also be tested properly on systemd and on cg v2.

/hold

odinuge commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -669,53 +591,21 @@ func (m *cgroupManagerImpl) ReduceCPULimits(cgroupName CgroupName) error {
-	var stats *libcontainercgroups.Stats
+// MemoryUsage returns the current memory usage of the specified cgroup,
+// as read from cgroupfs.
+func (m *cgroupManagerImpl) MemoryUsage(name CgroupName) (int64, error) {
```
Do you know of a test signal that can verify this change works as expected?

kolyshkin replied:
The code that is using this was added by #41149, and later moved out of experimental in #62509. Neither comes with any tests, but maybe @sjenning can shed more light on it (he should review this anyway).

fromanirh replied:
/cc @sjenning 
at least according to github UI it seems he was not assigned (just mentioned)

odinuge commented on pkg/kubelet/cm/types.go:
```diff
@@ -90,8 +78,8 @@ type CgroupManager interface {
 	ReduceCPULimits(cgroupName CgroupName) error
-	// GetResourceStats returns statistics of the specified cgroup as read from the cgroup fs.
-	GetResourceStats(name CgroupName) (*ResourceStats, error)
+	// MemoryUsage returns current memory usage of the specified cgroup, as read from the cgroupfs.
```
Maybe specify that it is in bytes and not pages/kB/MB?

kolyshkin replied:
Done. I have also removed that the value is read from cgroupfs as this is kind of obvious.

odinuge replied:
This still reads `MemoryUsage returns current memory usage of the specified cgroup, as read from the cgroupfs.` tho.

odinuge commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

odinuge commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

odinuge commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

kolyshkin commented:
```
E2eNode Suite: [sig-node] Summary API [NodeConformance] when querying /stats/summary should report resource usage through the stats api expand_less	1m55s
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/summary_test.go:54
Timed out after 90.000s.
Expected
    <string>: Summary
to match fields: {
[.Node.SystemContainers[runtime].Memory:
	Expected
	    <string>: MemoryStats
	to match fields: {
	.AvailableBytes:
		Expected
		    <*uint64 | 0xc000a91418>: 3705806848
		to be nil
	}
	, .Node.SystemContainers[pods].Memory:
	Expected
	    <string>: MemoryStats
	to match fields: {
	.MajorPageFaults:
		Expected
		    <uint64>: 198
		to be <=
		    <int>: 10
	}
	, .Node.SystemContainers[kubelet].Memory:
	Expected
	    <string>: MemoryStats
	to match fields: {
	.AvailableBytes:
		Expected
		    <*uint64 | 0xc000a91560>: 3836538880
		to be nil
	}
	]
}
```

As far as I understand, there are two issues here.

1. The test expects `MemoryStats.AvailableBytes` to be `nil` for system containers, and yet it reports something close to 4GB for `runtime` and `kubelet` (but not for `pods`).
2. Higher than expected number of `MemoryStats.MajorPageFaults` from the `pods` system container.

This is _presumably_ caused by the first commit (removing `propagateControllers`), and I can't think about why is that. Perhaps somehow writing to `cgroup.subtree_control` reset the state of the controller?

Looking at it.

odinuge commented:
> As far as I understand, there are two issues here.

Ahh, sorry, thought you knew about these test failures... They are expected, I should probably have stated that explicitly. The failures in `pull-kubernetes-node-crio-cgrpv2-e2e ` will be fixed by bumping cadvisor and the commit changing the conformance test in your PR. They are tracked here: https://github.com/kubernetes/kubernetes/issues/99230

`pull-kubernetes-node-kubelet-serial-crio-cgroupv{1,2}` are just experimental, so don't care about those... :sweat_smile: 


Overall I think this PR is ok. There are some issues with controller propagation on v2 in general (some WIP stuff here: https://github.com/kubernetes/kubernetes/pull/102250), but you know more about the runc part than me.

kolyshkin commented:
So, this PR can be merged then I guess?

fromanirh commented:
> > So, this PR can be merged then I guess?
> 
> Overall I think it is ok. I just need to think about why the `propagateControllers` was added in the first place. When dealing with systemd (both using systemd driver and procfs driver), systemd will manage controllers and stuff for you. It might be a case where `propagateControllers` had to be run before "update", but in a case where we don't have tests. The whole propagate controllers unless we are in a systemd delegated subtree is quite dodgy. I'll defer that to @giuseppe.
> 
> We don't have any testing of using cgroupfs on cgroup v2, and I am pretty sure running that while using systemd _will_ break, no matter what... :/
> 
> /cc @giuseppe

my 2c: I had a review as well and I fully agree with @odinuge , from my POV all the changes in this PR make sense, with the caveat above about `propagateControllers`. From my (admittedly quick) review of the history, I'm not sure this was not just a accidental duplicate of runc features, but better indeed wait for review from. @giuseppe

giuseppe commented:
> ```
> 
> ```
> 
> 
>     
>   
> 
> As far as I understand, there are two issues here.
> 
> 
>     1. The test expects `MemoryStats.AvailableBytes` to be `nil` for system containers, and yet it reports something close to 4GB for `runtime` and `kubelet` (but not for `pods`).
> 
> 
>     2. Higher than expected number of `MemoryStats.MajorPageFaults` from the `pods` system container.
> 
> 
> 
> 
> This is _presumably_ caused by the first commit (removing `propagateControllers`), and I can't think about why is that. Perhaps somehow writing to `cgroup.subtree_control` reset the state of the controller?

I forgot what tests was failing but there are two fixes in cAdvisor that AFAICS are not yet propagated into a release and into Kubernetes:

https://github.com/google/cadvisor/pull/2837
https://github.com/google/cadvisor/pull/2839

Especially the first one, solves an issue when reading memory stats

giuseppe commented:
> Overall I think it is ok. I just need to think about why the `propagateControllers` was added in the first place. When dealing with systemd (both using systemd driver and procfs driver), systemd will manage controllers and stuff for you. It might be a case where `propagateControllers` had to be run before "update", but in a case where we don't have tests. The whole propagate controllers unless we are in a systemd delegated subtree is quite dodgy. I'll defer that to @giuseppe.

at the time I've added that code, runc didn't work with cgroupfs on cgroup v2.  Now that this functionality is in libcontainer, I agree it is better to remove it

kolyshkin commented:
> > ```
> > 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As far as I understand, there are two issues here.
> > ```
> > 1. The test expects `MemoryStats.AvailableBytes` to be `nil` for system containers, and yet it reports something close to 4GB for `runtime` and `kubelet` (but not for `pods`).
> > 
> > 
> > 2. Higher than expected number of `MemoryStats.MajorPageFaults` from the `pods` system container.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is _presumably_ caused by the first commit (removing `propagateControllers`), and I can't think about why is that. Perhaps somehow writing to `cgroup.subtree_control` reset the state of the controller?
> 
> I forgot what tests was failing but there are two fixes in cAdvisor that AFAICS are not yet propagated into a release and into Kubernetes:
> 
> [google/cadvisor#2837](https://github.com/google/cadvisor/pull/2837)
> [google/cadvisor#2839](https://github.com/google/cadvisor/pull/2839)
> 
> Especially the first one, solves an issue when reading memory stats

Thanks, opened https://github.com/google/cadvisor/issues/2878 and https://github.com/google/cadvisor/pull/2879 (it might need a consent from you to make a google-cla bot happy).

matthyx commented:
/uncc

ehashman commented:
/assign @odinuge @fromanirh 

/cc @sjenning

odinuge reviewed (approved):
/lgtm

giuseppe reviewed (approved):
LGTM

odinuge commented:
/hold cancel
